src/cdef_apply.rs

```diff
@@ -1,270 +1,226 @@
 use crate::include::common::bitdepth::BitDepth;
+use crate::include::common::bitdepth::BPC;
 use crate::include::common::intops::ulog2;
 use crate::include::dav1d::headers::Rav1dPixelLayout;
 use crate::src::align::Align16;
 use crate::src::cdef::CdefEdgeFlags;
-use crate::src::cdef::CDEF_HAVE_BOTTOM;
-use crate::src::cdef::CDEF_HAVE_LEFT;
-use crate::src::cdef::CDEF_HAVE_RIGHT;
-use crate::src::cdef::CDEF_HAVE_TOP;
 use crate::src::internal::Rav1dContext;
 use crate::src::internal::Rav1dDSPContext;
 use crate::src::internal::Rav1dFrameContext;
 use crate::src::internal::Rav1dTaskContext;
 use crate::src::lf_mask::Av1Filter;
+use bitflags::bitflags;
 use libc::ptrdiff_t;
 use std::cmp;
 use std::ffi::c_int;
 use std::ffi::c_uint;
 use std::slice;
 
-pub type Backup2x8Flags = c_uint;
-pub const BACKUP_2X8_UV: Backup2x8Flags = 2;
-pub const BACKUP_2X8_Y: Backup2x8Flags = 1;
+bitflags! {
+    #[derive(Clone, Copy)]
+    struct Backup2x8Flags: u8 {
+        const Y = 1 << 0;
+        const UV = 1 << 1;
+    }
+}
+
+impl Backup2x8Flags {
+    pub const fn select(&self, select: bool) -> Self {
+        if select {
+            *self
+        } else {
+            Self::empty()
+        }
+    }
+}
 
 unsafe fn backup2lines<BD: BitDepth>(
-    dst: *const *mut BD::Pixel,
-    src: *const *mut BD::Pixel,
-    stride: *const ptrdiff_t,
+    dst: &[*mut BD::Pixel],
+    src: &[*mut BD::Pixel; 3],
+    stride: &[ptrdiff_t; 2],
     layout: Rav1dPixelLayout,
 ) {
-    let y_stride: ptrdiff_t = BD::pxstride(*stride.offset(0) as usize) as isize;
+    let y_stride: ptrdiff_t = BD::pxstride(stride[0] as usize) as isize;
     if y_stride < 0 {
+        let len = (-2 * y_stride) as usize;
```

Comment:
You could use `let len = 2 * y_stride.unsigned_abs();` for both `len`s instead.

Reply:
yeah that's nice, makes the similarity between both branches a bit clearer.